### PR TITLE
Multiline JSON support

### PR DESCRIPTION
**Pull request by eisber**
@lhoang29 review time :)

I split VowpalWabbitJsonSerializer into VowpalWabbitJsonSerializer  and VowpalWabbitJsonBuilder

### CODE REVIEW

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+                                if (reader.TokenType == JsonToken.StartObject)
+                                    this.ParseNamespaceAndFeatures(this.Context, propertyName);
+                                else
+                                    this.ParseFeature(this.DefaultNamespaceContext, defaultNamespace, propertyName);
+
+                            }
+                            break;
+                        case JsonToken.EndObject:
+                            goto done;
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: wouldn't break be enough here?

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                }
+            }
+        }
+
+        private void ParseSpecialProperty(VowpalWabbitMarshalContext context, Namespace ns, string propertyName)
+        {
+            // special fields
+            switch (propertyName)
+            {
+                case "_label":
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: didn't you already define this string as constant somewhere?

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        private void ParseFeatureArray(VowpalWabbitMarshalContext context, Namespace ns)
+        {
+            ulong index = 0;
+
+            while (reader.Read())
+            {
+                switch (reader.TokenType)
+                {
+                    case JsonToken.Integer:
+                        MarshalFloatFeature(context, ns, index, (float)(long)reader.Value);
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: why double cast?

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+                        break;
+                    case JsonToken.Float:
+                        MarshalFloatFeature(context, ns, index, (float)(double)reader.Value);
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: same here

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 5]
```diff
@@ -24,33 +24,270 @@ namespace VW.Serializer
...
+            /// </summary>
+            public abstract void Dispose();
+        }
 
         /// <summary>
-        /// Mapping from properties to types for labels.
+        /// Result for a single example.
         /// </summary>
-        private static readonly Dictionary<string, Type> labelPropertyMapping;
+        public sealed class SingleVowpalWabbitExample : VowpalWabbitExamples
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: perhaps rename this to SingleLine... to be in parity with Multiline

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 6]
```diff
@@ -24,33 +24,270 @@ namespace VW.Serializer
     /// </summary>
     public sealed class VowpalWabbitJsonSerializer : IDisposable
     {
-        internal const string FeatureIgnorePrefix = "_";
+        /// <summary>
+        /// Base class for JSON deserialization result.
+        /// </summary>
+        public abstract class VowpalWabbitExamples : IDisposable
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: hmm, this class names seems slightly confusing given that there's already one called VowpalWabbitExample. I don't have a better name in mind though

**Comment by lhoang29** [comment]
looks good in general

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs (hunk 1 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: We’re inside a switch. Break only breaks the switch, not the loop.

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs (hunk 2 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: No and done.

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs (hunk 3 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: reader.Value is of type object, the actual object is a long. Cast directly from object to float fails. The first cast gets us from object to long, the second actually does a conversion.

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 6 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: I was unhappy too. Renamed to VowpalWabbitExampleCollection and moved to separate files.

File: cs/cs/Serializer/VowpalWabbitJsonBuilder.cs (hunk 1 above)
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonBuilder.cs]: ah, duh

**Pull request merged**